src/timestamp_generator.hpp

```diff
@@ -61,16 +61,24 @@ class ServerSideTimestampGenerator : public TimestampGenerator {
 
 class MonotonicTimestampGenerator : public TimestampGenerator {
 public:
-  MonotonicTimestampGenerator()
+  MonotonicTimestampGenerator(int64_t warning_threshold_us = 1000000,
+                              int64_t warning_interval_ms = 1000)
     : TimestampGenerator(MONOTONIC)
-    , last_(0) { }
+    , last_(0)
+    , last_warning_(0)
+    , warning_threshold_us_(warning_threshold_us)
+    , warning_interval_ms_(warning_interval_ms) { }
 
   virtual int64_t next();
 
 private:
   int64_t compute_next(int64_t last);
 
   Atomic<int64_t> last_;
+  Atomic<int64_t> last_warning_;
+
+  const int64_t warning_threshold_us_;
+  const int64_t warning_interval_ms_;
```

Comment:
Interesting that the compiler doesn't warn when it knows you're trying to put a negative value into an unsigned type.

Also, why do you think it was a mistake to expose unsigned values? It shows an intent for valid argument values (even if the compiler doesn't enforce it).